contracts/ERC721Collateralizer.sol

```diff
@@ -0,0 +1,426 @@
+
+    function ERC721Collateralizer(
+        address _contractRegistry,
+        address _tokenRegistry,
+        address _cryptoKittiesContract
+    ) public {
+        ContractRegistry contractRegistry = ContractRegistry(_contractRegistry);
+
+        // Get the DebtRegistry contract from the ContractRegistry.
+        debtRegistry = contractRegistry.debtRegistry();
+        // Instantiate a special token registry for ERC721s.
+        tokenRegistry = ERC721TokenRegistry(_tokenRegistry);
+        // Keep track of the CryptoKitties contract, for specific workarounds.
+        cryptoKittiesContract = _cryptoKittiesContract;
+    }
+
+    /**
+     * Transfers collateral from the debtor to the current contract, as custodian.
+     *
+     * @param agreementId bytes32 The debt agreement's ID
+     * @param debtor address The owner of the asset being collateralized
+     */
+    function collateralize(
+        bytes32 agreementId,
+        address debtor
+    )
+        public
+        onlyAuthorizedToCollateralize
+        whenNotPaused
+        returns (bool _success)
+    {
+        // The address of the ERC721 contract.
+        address collateralTokenAddress;
+        // The token id, associating the collateral token and the given ERC721 contract.
+        uint256 collateralTokenID;
+        // The terms contract according to which this asset is being collateralized.
+        TermsContract termsContract;
+
+        // Fetch all relevant collateralization parameters
+        (
+            collateralTokenAddress,
+            collateralTokenID,
+            termsContract
+        ) = retrieveCollateralParameters(agreementId);
+
+        require(termsContract == msg.sender);
+        require(collateralTokenAddress != address(0));
+
+        /*
+            Ensure that the agreement has not already been collateralized.
+
+            If the agreement has already been collateralized, this check will fail
+            because any valid collateralization must have some sort of valid
+            address associated with it as a debtor.  Given that it is impossible
+            to send transactions from address 0x0, this check will only fail
+            when the agreement is already collateralized.
+        */
+        require(agreementToDebtor[agreementId] == address(0));
+
+        ERC721 erc721token = ERC721(collateralTokenAddress);
+        address custodian = address(this);
+
+        // Ensure that the debtor is the owner of this token ID.
+        require(erc721token.ownerOf(collateralTokenID) == debtor);
+
+        // Transfer the collateral asset to this contract.
+        erc721token.transferFrom(debtor, custodian, collateralTokenID);
+
+        // Ensure that this contract is in fact the owner of the collateral.
+        require(erc721token.ownerOf(collateralTokenID) == custodian);
+
+        // Store debtor in mapping, effectively demarcating that the agreement is now collateralized.
+        agreementToDebtor[agreementId] = debtor;
```

Comment:
Nit: re-entrancy smell.  Pretty sure there's not a vulnerability here, but I think it's generally a good idea to do the sorts of declarations that demarcate a certain action on an external contract no-longer do-able _before_ that action is done -- i.e. this should go immediately before the `transferFrom` call.  See https://medium.com/@gus_tavo_guim/reentrancy-attack-on-smart-contracts-how-to-identify-the-exploitable-and-an-example-of-an-attack-4470a2d8dfe4